### PR TITLE
Current Vision Code

### PR DESCRIPTION
**Pull request by Injector-Spenral**

### CODE REVIEW

File: cameraServer.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+minContourArea = 10
+angleOffset = 13
+rightAngleSize = -14.5
+leftAngleSize = -75.5
+screenSize = (320, 240)
+
+#Initialisation
+configFile = "/boot/frc.json"
+
+class CameraConfig: pass
```
Comment (fizzlebert) [inline cameraServer.py]: Why is this class created if it has not body.

File: cameraServer.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+cameraConfigs = []
+
+"""Read single camera configuration."""
+def readCameraConfig(config):
+    cam = CameraConfig()
+
+    # name
+    try:
+        cam.name = config["name"]
+    except KeyError:
```
Comment (fizzlebert) [inline cameraServer.py]: Why are there so many try except statements, if you are that worried about keys not existing try
```python
for key in (key1, key2):
    if key not in config:
        # error handling
```

File: cameraServer.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            return angle, -(((x/screenSize[0])*2)-1), img
+    return None, None, img
+
+def getGroundPos(img, sample=False): #TODO: Ground Tape Function
+    return None, None, None #Distance from x, distance from y, angle of line
+    
+#if False:
+if __name__ == "__main__":
+    from cscore import CameraServer, VideoSource
+    from networktables import NetworkTables
```
Comment (fizzlebert) [inline cameraServer.py]: Read [pep 8](https://pep8.org/#imports) for how to structure imports, hint, should go at the top of the file

File: cameraServer.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    except KeyError:
+        parseError("camera '{}': could not read path".format(cam.name))
+        return False
+
+    cam.config = config
+
+    cameraConfigs.append(cam)
+    return True
+
+"""Report parse error."""
```
Comment (fizzlebert) [inline cameraServer.py]: Docstrings go underneath a function

File: cameraServer.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+def parseError(str):
```
Comment (fizzlebert) [inline cameraServer.py]: [str](https://docs.python.org/3/library/functions.html#func-str) is a built in and should not be used as a function argument

File: cameraServer.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    # team number
+    try:
+        team = j["team"]
+    except KeyError:
+        parseError("could not read team number")
+        return False
+
+    # ntmode (optional)
+    if "ntmode" in j:
+        str = j["ntmode"]
```
Comment (fizzlebert) [inline cameraServer.py]: Again str is a built in function not to be used as a variable name

File: cameraServer.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    # parse file
+    try:
+        with open(configFile, "rt") as f:
+            j = json.load(f)
+    except OSError as err:
+        print("could not open '{}': {}".format(configFile, err), file=sys.stderr)
+        return False
+
+    # top level must be an object
+    if not isinstance(j, dict):
```
Comment (fizzlebert) [inline cameraServer.py]: Would this ever be true :question:

File: cameraServer.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+    # ntmode (optional)
+    if "ntmode" in j:
+        str = j["ntmode"]
+        if str.lower() == "client":
+            server = False
+        elif str.lower() == "server":
+            server = True
+        else:
+            parseError("could not understand ntmode value '{}'".format(str))
```
Comment (fizzlebert) [inline cameraServer.py]: Again, f-string save characters

File: cameraServer.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        return False
+    for camera in cameras:
+        if not readCameraConfig(camera):
+            return False
+
+    return True
+
+"""Start running the camera."""
+def startCamera(config):
+    print("Starting camera '{}' on {}".format(config.name, config.path))
```
Comment (fizzlebert) [inline cameraServer.py]: F-string

File: cameraServer.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    sink = cameras[0][0].getVideo(camera=cameras[0][1])
+    source = cameras[0][0].putVideo('cv', 320, 240)
+
+    frame = np.zeros(shape=(screenSize[1], screenSize[0], 3))
+    while True:
+        time1, frame = sink.grabFrameNoTimeout(image=frame)
+        angle, percent, image = getRetroPos(frame, True)
+
+        if not percent:
+            percent = 999
```
Comment (fizzlebert) [inline cameraServer.py]: Sending a real number when no angle exists is not ideal behaviour try `float("NaN")` instead

File: cameraServer.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        circleContours = list(np.int0(cv2.boxPoints(closestToMiddle[0])))
+        circleContours.extend(list(np.int0(cv2.boxPoints(closestToMiddle[1]))))
+        circleContours = np.array(circleContours)
+        (x,y),radius = cv2.minEnclosingCircle(circleContours)
+        radius = int(radius)
+        center = (int(x), int(y))
+
+        angle = (closestToMiddle[0][1][0] * closestToMiddle[0][1][1]) / (closestToMiddle[1][1][0] * closestToMiddle[1][1][1])
+        
+        if display == True: #Create the annotated display if display is True
```
Comment (fizzlebert) [inline cameraServer.py]: Save character use [truth value testing](https://docs.python.org/3/library/stdtypes.html#truth-value-testing)
```suggestion
        if display: #Create the annotated display if display is True
```

File: cameraServer.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+import time
+import sys
+import numpy as np
+import cv2
+import math
+
+#Magic Numbers
+lowerGreen = (38, 125, 100) #Our Robot's Camera
+higherGreen = (52, 255, 180)
+sample_lowerGreen = (30, 177, 80) #FRC sample images
```
Comment (fizzlebert) [inline cameraServer.py]: What variable naming convention are you using, [Camel Case](https://en.wikipedia.org/wiki/Camel_case) for two lines then [Snake Case](https://en.wikipedia.org/wiki/Snake_case) for two lines :laughing:, decide on one

**Review by fizzlebert — CHANGES_REQUESTED** [review]

**Comment by auscompgeek** [comment]
@danieloconell complains about WPILib's example code should go to WPILib, not here :)

**Comment by auscompgeek** [comment]
It is very unclear what is our code and what is not. Please make a clear distinction in the commit log.

File: cameraServer.py (hunk 8 above)
Comment (fizzlebert) [reply cameraServer.py]: And f-strings don't work in python 3.5

File: cameraServer.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+    print("config error in '" + configFile + "': " + error_str, file=sys.stderr)
+
+def readConfig():
+    """Read configuration file."""
+    global team
+    global server
+
+    # parse file
+    try:
+        with open(configFile, "rt") as f:
```
Comment (fizzlebert) [inline cameraServer.py]: By default [open](https://docs.python.org/3/library/functions.html#open) uses `r` and `t` when opening a file
```suggestion
        with open(configFile) as f:
```

Comment (auscompgeek) [reply cameraServer.py]: Again, this is not the place to complain about the example code.

**Pull request merged**